src/com/xilinx/rapidwright/eco/ECOTools.java

```diff
@@ -1062,6 +1106,68 @@ public static void createCell(Design design, EDIFCell reference, List<String> pa
+     * instances are not possible.
+     * 
+     * @param design    The existing design
+     * @param input     The reference logical input pin on which the new inline cell
+     *                  should be created and connected.
+     * @param reference The primitive cell type to create.
+     * @param site      The site on which the new cell should be placed.
+     * @param bel       The BEL within the provided site onto which the cell should
+     *                  be placed.
+     * @param logInput  The logical cell's input pin that should connect to the
+     *                  existing input's net
+     * @param logOutput The logical cell's output pin that should connect to the
+     *                  existing input
+     * @return A hierarchical reference to the newly created cell instance.
+     */
+    public static Cell createInlineCellOnInputPin(Design design, EDIFHierPortInst input,
+            Unisim reference, Site site, BEL bel, String logInput, String logOutput) {
+        if (!input.isInput()) return null;
+        EDIFHierNet net = input.getHierarchicalNet();
+        EDIFCell parent = input.getPortInst().getParentCell();
+        EDIFNet newNet = parent.createNet("inline_insertion_net_" + UNIQUE_COUNT++);
+        EDIFHierNet newHierNet = new EDIFHierNet(net.getHierarchicalInst(), newNet);
+
+        // Create new cell instance, add to primitives library if necessary
+        EDIFCell refCell = design.getNetlist().getHDIPrimitivesLibrary().getCell(reference.name());
+        if (refCell == null) {
+            refCell = Design.getUnisimCell(reference);
+            design.getNetlist().getHDIPrimitivesLibrary().addCell(refCell);
+        }
+        EDIFCellInst newInst = refCell.createCellInst("inline_insertion_" + refCell.getName()
+                + "_" + UNIQUE_COUNT++, parent);
+        EDIFHierCellInst newHierInst = net.getHierarchicalInst().getChild(newInst);
+        Cell cell = design.createCell(newHierInst.getFullHierarchicalInstName(), newInst);
+        if (!design.placeCell(cell, site, bel)) {
+            throw new RuntimeException("ERROR: Unable to create new inline cell to be placed on " 
+                + site + "/" + bel);
+        }
+
+        EDIFPortInst newInput = new EDIFPortInst(refCell.getPort(logInput), null, newInst);
+        EDIFPortInst newOutput = new EDIFPortInst(refCell.getPort(logOutput), null, newInst);
+        EDIFHierPortInst newHierInput = newHierInst.getPortInst(newInput.getName());
+        EDIFHierPortInst newHierOutput = newHierInst.getPortInst(newOutput.getName());
+        
+        disconnectNet(design, input);
+        
+        // Connect new cell input to existing net, output and existing input to new net
+        Map<EDIFHierNet, List<EDIFHierPortInst>> connectMap = new HashMap<>();
+        connectMap.put(net, new ArrayList<>(Arrays.asList(newHierInput)));
+        connectMap.put(newHierNet, new ArrayList<>(Arrays.asList(newHierOutput, input)));
```

Comment:
Is a copy necessary here? `connectMap` should not be mutating the list.
```suggestion
        connectMap.put(net, new ArrayList<>(Arrays.asList(newHierInput)));
        connectMap.put(newHierNet, new ArrayList<>(Arrays.asList(newHierOutput, input)));
```

Reply:
If I apply this change, it fails:
![image](https://github.com/Xilinx/RapidWright/assets/20803522/5323b57f-ab96-44f4-8edc-5d6b3ddb6666)